graphsync/main.go

```diff
@@ -0,0 +1,340 @@
+	defer initCtx.SyncClient.MustSignalAndWait(ctx, "done", runenv.TestInstanceCount)
+
+	host, peers := makeHost(ctx, runenv, initCtx)
+	defer host.Close()
+
+	var (
+		// make datastore, blockstore, dag service, graphsync
+		ds     = dss.MutexWrap(datastore.NewMapDatastore())
+		bs     = blockstore.NewBlockstore(ds)
+		dagsrv = merkledag.NewDAGService(blockservice.New(bs, offline.Exchange(bs)))
+		gsync  = gsi.New(ctx, gsnet.NewFromLibp2pHost(host), makeLoader(bs), makeStorer(bs))
+	)
+
+	switch runenv.TestGroupID {
+	case "providers":
+		if runenv.TestGroupInstanceCount > 1 {
+			panic("test case only supports one provider")
+		}
+
+		runenv.RecordMessage("we are the provider")
+		defer runenv.RecordMessage("done provider")
+
+		return runProvider(ctx, runenv, initCtx, dagsrv, size, latencies, bandwidths)
+
+	case "requestors":
+		runenv.RecordMessage("we are the requestor")
+		defer runenv.RecordMessage("done requestor")
+
+		p := *peers[0]
+		if err := host.Connect(ctx, p); err != nil {
+			return err
+		}
+		runenv.RecordMessage("done dialling provider")
+		return runRequestor(ctx, runenv, initCtx, gsync, p, bs, latencies, bandwidths)
+
+	default:
+		panic("unsupported group ID")
+	}
+}
+
+func runRequestor(ctx context.Context, runenv *runtime.RunEnv, initCtx *run.InitContext, gsync gs.GraphExchange, p peer.AddrInfo, bs blockstore.Blockstore, latencies []time.Duration, bandwidths []uint64) error {
+	// create a selector for the whole UnixFS dag
+	ssb := builder.NewSelectorSpecBuilder(basicnode.Style.Any)
+	sel := ssb.ExploreRecursive(
+		selector.RecursionLimitNone(),
+		ssb.ExploreAll(
+			ssb.ExploreRecursiveEdge()),
+	).Node()
+
+	for i, latency := range latencies {
+		for j, bandwidth := range bandwidths {
+			round := i*len(latencies) + j
+
+			var (
+				topicCid  = sync.NewTopic(fmt.Sprintf("cid-%d", round), new(cid.Cid))
+				stateNext = sync.State(fmt.Sprintf("next-%d", round))
+				stateNet  = sync.State(fmt.Sprintf("network-configured-%d", round))
+			)
+
+			runenv.RecordMessage("waiting to start round %d", round)
+			initCtx.SyncClient.MustSignalAndWait(ctx, stateNext, runenv.TestInstanceCount)
+
+			sctx, scancel := context.WithCancel(ctx)
+			cidCh := make(chan *cid.Cid, 1)
+			initCtx.SyncClient.MustSubscribe(sctx, topicCid, cidCh)
+			cid := <-cidCh
+			scancel()
+
+			// make a go-ipld-prime link for the root UnixFS node
+			clink := cidlink.Link{Cid: *cid}
+
+			runenv.RecordMessage("ROUND %d: latency=%s, bandwidth=%d", round, latency, bandwidth)
+			runenv.RecordMessage("CID: %s", cid)
+
+			runenv.RecordMessage("waiting for provider's network to be configured %d", round)
+			<-initCtx.SyncClient.MustBarrier(ctx, stateNet, 1).C
+			runenv.RecordMessage("network configured for round %d", round)
+
+			// execute the traversal.
+			runenv.RecordMessage(">>>>> requesting")
+			progressCh, errCh := gsync.Request(ctx, p.ID, clink, sel)
+			for r := range progressCh {
+				runenv.RecordMessage("******* progress: %+v", r)
+			}
+
```

Comment:
One thing you can do here is reread out the file to make sure you have all the bytes -- though you'd need to pass the original file bytes through with the CID in the channel (that might be a bit inefficient -- or maybe the processes can can both read the same system file? I'm not sure the constraints in testground for sharing large chunks of data)

You can see the code you can use to verify the bytes on the other side here: https://github.com/ipfs/go-graphsync/blob/master/impl/graphsync_test.go#L847-L865